kubelet/cm: fix bug where kubelet restarts from missing cpuset cgroup

haircommander opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug
#### What this PR does / why we need it:
on None cpumanager policy, cgroupv2, and systemd cgroup manager, kubelet could get into a situation where it believes the cpuset cgroup was created (by libcontainer in the cgroupfs) but systemd has deleted it, as it wasn't requested to create it. This causes one unnecessary restart, as kubelet fails with

`failed to initialize top level QOS containers: root container [kubepods] doesn't exist.`

This only causes one restart because the kubelet skips recreating the cgroup if it already exists, but it's still a bother and is fixed this way

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/122955

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fix a bug where the kubelet ephemerally fails with `failed to initialize top level QOS containers: root container [kubepods] doesn't exist`, due to the cpuset cgroup being deleted on v2 with systemd cgroup manager.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

haircommander committed:
kubelet/cm: fix bug where kubelet restarts from missing cpuset cgroup

on None cpumanager policy, cgroupv2, and systemd cgroup manager, kubelet
could get into a situation where it believes the cpuset cgroup was created
(by libcontainer in the cgroupfs) but systemd has deleted it, as it wasn't requested
to create it. This causes one unnecessary restart, as kubelet fails with

`failed to initialize top level QOS containers: root container [kubepods] doesn't exist.`

This only causes one restart because the kubelet skips recreating the cgroup
if it already exists, but it's still a bother and is fixed this way

Signed-off-by: Peter Hunt <pehunt@redhat.com>

kolyshkin commented on pkg/kubelet/cm/node_container_manager_linux.go:
```diff
@@ -183,9 +187,37 @@ func getCgroupConfig(rl v1.ResourceList) *ResourceConfig {
+	// By default, systemd will not create it, as we've not chosen to delegate it, and we haven't included it in the Apply() request.
+	// However, this causes a bug where kubelet restarts unnecessarily (cpuset cgroup is created in the cgroupfs, but systemd
+	// doesn't know about it and deletes it, and then kubelet doesn't continue because the cgroup isn't configured as expected).
+	// An alternative is to delegate the `cpuset` cgroup to the kubelet, but that would require some plumbing in libcontainer,
```
Just curious, what plumbing do you refer to here? IOW what's missing?

haircommander replied:
we currently use libcontainer for all of our cgroup management and there's no way to set a systemd property through the libcontainer manager. we could just use a godbus instance ourselves but it'd take some setup and copied code

kolyshkin commented on pkg/kubelet/cm/types.go:
```diff
@@ -19,12 +19,15 @@ package cm
 type ResourceConfig struct {
 	// Memory limit (in bytes).
 	Memory *int64
+	// CPU set (number of cpus the cgroup has access to)
```
nit: missing period

kolyshkin commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -131,6 +132,10 @@ type containerManagerImpl struct {
 	draManager dra.Manager
+	// The full set of CPUs on the node. This field is set lazily, and is used to make sure
+	// the `cpuset` cgroup hierarchy is created on cgroup v2 when cpumanager is using a
+	// None policy
```
nit: missing period.

kolyshkin commented on pkg/kubelet/cm/node_container_manager_linux.go:
```diff
@@ -183,9 +187,37 @@ func getCgroupConfig(rl v1.ResourceList) *ResourceConfig {
 }
 
+func (cm *containerManagerImpl) getAllCPUs() cpuset.CPUSet {
+	machineInfo, err := cm.cadvisorInterface.MachineInfo()
+	if err != nil {
+		klog.V(4).InfoS("Failed to get machine info to get default cpuset", "error", err)
+		return cpuset.CPUSet{}
+	}
+	topo, err := topology.Discover(machineInfo)
+	if err != nil {
+		klog.V(4).InfoS("Failed to get topology info to get default cpuset", "error", err)
+		return cpuset.CPUSet{}
+	}
+	return topo.CPUDetails.CPUs()
```
A weird idea: what if we just take the contents of `/sys/fs/cgroup/cpuset.cpus.effective`?

haircommander replied:
that's an option. I opted for this because it matches what the cpumanager does to initialize the full set of CPUs, which i figure may be more consistent to have the kubelet gather the cpu list one way. I am open though.

ffromani commented:
/triage accepted
/priority important-longterm

ffromani commented on pkg/kubelet/cm/node_container_manager_linux.go:
```diff
@@ -183,9 +187,37 @@ func getCgroupConfig(rl v1.ResourceList) *ResourceConfig {
 	return &rc
 }
 
+func (cm *containerManagerImpl) getAllCPUs() cpuset.CPUSet {
```
just thinking aloud, are we in the container manager flow before the cpumanager is initialized?
I'd like to explore the option to put this logic inside cpumanager, so we can avoid to peek on its options from the outside and to duplicate the topology discovery logic.
Note: I'm _NOT_ suggesting to pivot to this approach in this PR, just exploring (myself) the option to see how it looks.

ffromani replied:
ok, after a quick check we do have a `cpumanager` instance in the `containerManagerImpl` when we reach this code, so moving the functionality inside `cpumanager` and remove quite some duplication is at least possible.

ffromani replied:
sketch (utterly untested, for demo purposes only): 
```
diff --git a/pkg/kubelet/cm/cpumanager/cpu_manager.go b/pkg/kubelet/cm/cpumanager/cpu_manager.go
index 8b5049d7d74..e7fb1cdb8aa 100644
--- a/pkg/kubelet/cm/cpumanager/cpu_manager.go
+++ b/pkg/kubelet/cm/cpumanager/cpu_manager.go
@@ -93,6 +93,10 @@ type Manager interface {
 	// GetCPUAffinity returns cpuset which includes cpus from shared pools
 	// as well as exclusively allocated cpus
 	GetCPUAffinity(podUID, containerName string) cpuset.CPUSet
+
+	// GetAllCPUs returns all the CPUs known by cpumanager, as reported by the
+	// hardware discovery. Maps to the CPU capacity.
+	GetAllCPUs() cpuset.CPUSet
 }
 
 type manager struct {
@@ -136,7 +140,11 @@ type manager struct {
 	// stateFileDirectory holds the directory where the state file for checkpoints is held.
 	stateFileDirectory string
 
-	// allocatableCPUs is the set of online CPUs as reported by the system
+	// allCPUs is the set of online CPUs as reported by the system
+	allCPUs cpuset.CPUSet
+
+	// allocatableCPUs is the set of online CPUs as reported by the system,
+	// and available for allocation, minus the reserved set
 	allocatableCPUs cpuset.CPUSet
 
 	// pendingAdmissionPod contain the pod during the admission phase
@@ -156,6 +164,11 @@ func NewManager(cpuPolicyName string, cpuPolicyOptions map[string]string, reconc
 	var policy Policy
 	var err error
 
+	topo, err = topology.Discover(machineInfo)
+	if err != nil {
+		return nil, err
+	}
+
 	switch policyName(cpuPolicyName) {
 
 	case PolicyNone:
@@ -165,10 +178,6 @@ func NewManager(cpuPolicyName string, cpuPolicyOptions map[string]string, reconc
 		}
 
 	case PolicyStatic:
-		topo, err = topology.Discover(machineInfo)
-		if err != nil {
-			return nil, err
-		}
 		klog.InfoS("Detected CPU topology", "topology", topo)
 
 		reservedCPUs, ok := nodeAllocatableReservation[v1.ResourceCPU]
@@ -205,6 +214,7 @@ func NewManager(cpuPolicyName string, cpuPolicyOptions map[string]string, reconc
 		topology:                   topo,
 		nodeAllocatableReservation: nodeAllocatableReservation,
 		stateFileDirectory:         stateFileDirectory,
+		allCPUs:                    topo.CPUDetails().CPUs(),
 	}
 	manager.sourcesReady = &sourcesReadyStub{}
 	return manager, nil
@@ -339,6 +349,10 @@ func (m *manager) GetAllocatableCPUs() cpuset.CPUSet {
 	return m.allocatableCPUs.Clone()
 }
 
+func (m *manager) GetAllCPUs() cpuset.CPUSet {
+	return m.allCPUs.Clone()
+}
+
 type reconciledContainer struct {
 	podName       string
 	containerName string
diff --git a/pkg/kubelet/cm/node_container_manager_linux.go b/pkg/kubelet/cm/node_container_manager_linux.go
index 9c9c91bc6f2..42d5f8939e3 100644
--- a/pkg/kubelet/cm/node_container_manager_linux.go
+++ b/pkg/kubelet/cm/node_container_manager_linux.go
@@ -31,12 +31,9 @@ import (
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	"k8s.io/klog/v2"
 	kubefeatures "k8s.io/kubernetes/pkg/features"
-	"k8s.io/kubernetes/pkg/kubelet/cm/cpumanager"
-	"k8s.io/kubernetes/pkg/kubelet/cm/cpumanager/topology"
 	"k8s.io/kubernetes/pkg/kubelet/events"
 	"k8s.io/kubernetes/pkg/kubelet/stats/pidlimit"
 	kubetypes "k8s.io/kubernetes/pkg/kubelet/types"
-	"k8s.io/utils/cpuset"
 )
 
 const (
@@ -194,9 +191,10 @@ func (cm *containerManagerImpl) getCgroupConfig(rl v1.ResourceList) *ResourceCon
 	// An alternative is to delegate the `cpuset` cgroup to the kubelet, but that would require some plumbing in libcontainer,
 	// and this is sufficient.
 	// Only do so on None policy, as Static policy will do its own updating of the cpuset.
-	if cm.NodeConfig.CPUManagerPolicy == string(cpumanager.PolicyNone) {
+	// Please see the comment on policy none's GetAllocatableCPUs
+	if cm.cpuManager.GetAllocatableCPUs().IsEmpty() {
 		if cm.allCPUs.IsEmpty() {
-			cm.allCPUs = cm.getAllCPUs()
+			cm.allCPUs = cm.cpuManager.GetAllCPUs()
 		}
 		rc.CPUSet = cm.allCPUs
 	}
@@ -204,20 +202,6 @@ func (cm *containerManagerImpl) getCgroupConfig(rl v1.ResourceList) *ResourceCon
 	return &rc
 }
 
-func (cm *containerManagerImpl) getAllCPUs() cpuset.CPUSet {
-	machineInfo, err := cm.cadvisorInterface.MachineInfo()
-	if err != nil {
-		klog.V(4).InfoS("Failed to get machine info to get default cpuset", "error", err)
-		return cpuset.CPUSet{}
-	}
-	topo, err := topology.Discover(machineInfo)
-	if err != nil {
-		klog.V(4).InfoS("Failed to get topology info to get default cpuset", "error", err)
-		return cpuset.CPUSet{}
-	}
-	return topo.CPUDetails.CPUs()
-}
-
 // GetNodeAllocatableAbsolute returns the absolute value of Node Allocatable which is primarily useful for enforcement.
 // Note that not all resources that are available on the node are included in the returned list of resources.
 // Returns a ResourceList.
```

haircommander replied:
yeah this is better I've pushed an adapted version, PTAL

haircommander committed:
kubelet/cm: move CPU reading from cm to cm/cpumanager

Authored-by: Francesco Romani <fromani@redhat.com>
Signed-off-by: Peter Hunt <pehunt@redhat.com>

kwilczynski commented on pkg/kubelet/cm/types.go:
```diff
@@ -19,12 +19,15 @@ package cm
 type ResourceConfig struct {
 	// Memory limit (in bytes).
 	Memory *int64
+	// CPU set (number of cpus the cgroup has access to).
```
Nit.

```suggestion
	// CPU set (number of CPUs the cgroup has access to).
```

For consistency.

haircommander replied:
fixed!

kwilczynski commented:
/approve 
/lgtm

mrunalp commented:
@ffromani Can you make a pass before I approve?

ffromani commented on pkg/kubelet/cm/cgroup_manager_linux.go:
```diff
@@ -298,6 +298,9 @@ func (m *cgroupCommon) toResources(resourceConfig *ResourceConfig) *libcontainer
 	if resourceConfig.PidsLimit != nil {
 		resources.PidsLimit = *resourceConfig.PidsLimit
 	}
+	if !resourceConfig.CPUSet.IsEmpty() {
+		resources.CpusetCpus = resourceConfig.CPUSet.String()
+	}
```
non-blocking question for better understanding: what would be the drawback, if any, in always setting the `CpusetCpus`?
this PR covers the `none` policy flow
the exclusively-allocated CPUs should be wanting to set this value anyway (we do in the hooks IIRC?)
the `static` policy without exclusively allocated CPUs should be not too hard.

haircommander replied:
mostly as a very targeted change because I don't immediately know the consequences of setting it in the other cases and I'd rather get the bug fix through

ffromani replied:
that's fair and reasonnable, let's keep it this way.

ffromani commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -339,6 +349,10 @@ func (m *manager) GetAllocatableCPUs() cpuset.CPUSet {
+func (m *manager) GetAllCPUs() cpuset.CPUSet {
+	return m.allCPUs.Clone()
+}
+
```
Thinking aloud (so totally not blocking) we will need some refactoring cleanup about available/allocatable/all/reserved cpusets in the future

ffromani reviewed (commented):
the implemenation looks reasonnable. Let's make sure we have tests coverage (e2e also) for this change; I'm mostly interested in the non-regression flow for starters.
Few non-blocking questions inside

ffromani commented:
/retest

the lane failure seems unrelated

ffromani commented:
@haircommander LGTM if we have test coverage, I strongly prefer e2e coverage, at least non-regression testing.

haircommander commented:
hey @ffromani I am struggling to write a test (even a unit) that verifies we can create a cgroup with the cgroup manager and then call Validate() on it and have that succeed. I will spend a bit more time on it but I think the value of this change is high enough that we may want to push through the fix without a test for now. WDYT?

ffromani commented:
> hey @ffromani I am struggling to write a test (even a unit) that verifies we can create a cgroup with the cgroup manager and then call Validate() on it and have that succeed. I will spend a bit more time on it but I think the value of this change is high enough that we may want to push through the fix without a test for now. WDYT?

I hear you, but how can we check at least non-regression?

haircommander commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

haircommander commented:
I'm gonna carry the second commit here https://github.com/kubernetes/kubernetes/pull/126602 . I am pretty sure it passes on master, which means it doesn't test this situation :cry: I'm about at my wits-end on catching the kubelet restart issue.

ffromani reviewed (commented):
@haircommander I'd like to try writing a e2e test to cover this flow if you don't mind. I want to try my hardest before to give up because this flow is pretty critical - `policy: none` is AFAIK the default. If I can't converge in a week or so we will re-evaluate.

ffromani committed:
e2e_node: add a test to verify the kubelet starts

with systemd cgroup driver and cpumanager none policy.

This was originally planned to be a correctness check for
https://issues.k8s.io/125923, but it was difficult to reproduce the bug,
so it's now a regression test against it.

Signed-off-by: Francesco Romani <fromani@redhat.com>
Signed-off-by: Peter Hunt <pehunt@redhat.com>

ffromani commented:
> @haircommander I'd like to try writing a e2e test to cover this flow if you don't mind. I want to try my hardest before to give up because this flow is pretty critical - `policy: none` is AFAIK the default. If I can't converge in a week or so we will re-evaluate.

contributed the e2e test which @haircommander graciosuly agreed to pick up. From my perspective this is the best we can do testwise, so my concerns are addressed (explained more details in the test proper)

I'll have a final pass before to add my formal LGTM.

ffromani reviewed (commented):
/lgtm

kwilczynski commented:
/approve
/lgtm

bart0sh commented:
/assign @SergeyKanzhelev @mrunalp 
for approval

SergeyKanzhelev commented on cmd/kubelet/app/server.go:
```diff
@@ -713,6 +713,8 @@ func run(ctx context.Context, s *options.KubeletServer, kubeDeps *kubelet.Depend
 		}
 	}
 
+	klog.InfoS("Using cgroup driver", "cgroupDriver", s.CgroupDriver)
```
is `KubeletCgroupDriverFromCRI` enabled, this will be double-logged. Maybe put it in `else` statement or remove it from https://github.com/kubernetes/kubernetes/blob/a7702cb47babcff60518b510293dbb41ba8eab35/cmd/kubelet/app/server.go#L1431

ffromani replied:
Another option would be bumping to `V(4)` or so. It was added mostly for debugging, it will indeed possibly double logged but exactly once, which should not be a big deal.
But also removing this line entirely is not a big deal at all.

ffromani replied:
it seems to me the log is still present in the last upload, can we please check again?

haircommander replied:
whoops, rebased wrong

haircommander commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/retest

SergeyKanzhelev commented:
> it believes the cpuset cgroup was created (by libcontainer in the cgroupfs) but systemd has deleted it, as it wasn't requested to create it.

is there any way to differentiate the wrong state and systemd delaying creation? I am not quite familiar with the logic and when systemd will create it. Will there be a possibility of a race condition if we are proceeding instead of keep restarting before it is created?

haircommander commented:
from my understanding the race happened because we created the cpuset cgroup without telling systemd we were (because there were no cpus in the cpuset), and thus systemd deleted it after it was created. adding the cpuset to the request will cause libcontainer to request systemd create it, and then it won't clobber it.

SergeyKanzhelev commented:
@ffromani lgtm?

ffromani commented:
/lgtm

sorry folks, this feel through the cracks. I think is good as we can make it, I'm happy with the change as-is now.

SergeyKanzhelev commented on test/e2e_node/node_container_manager_test.go:
```diff
@@ -76,6 +76,82 @@ var _ = SIGDescribe("Node Container Manager", framework.WithSerial(), func() {
+
+			ginkgo.DeferCleanup(func(ctx context.Context) {
+				if oldCfg != nil {
+					// Update the Kubelet configuration.
+					ginkgo.By("Stopping the kubelet")
+					startKubelet := stopKubelet()
+
+					// wait until the kubelet health check will fail
+					gomega.Eventually(ctx, func() bool {
+						return kubeletHealthCheck(kubeletHealthCheckURL)
+					}).WithTimeout(time.Minute).WithPolling(time.Second).Should(gomega.BeFalseBecause("expected kubelet health check to be failed"))
+					ginkgo.By("Stopped the kubelet")
+
+					framework.ExpectNoError(e2enodekubelet.WriteKubeletConfigFile(oldCfg))
```
is it the right ordering? Should we write file first and then restart? Presumably it will be faster

SergeyKanzhelev replied:
I mean you will not need to do separate stop and start and will just do `restartKubelet`.

SergeyKanzhelev replied:
We likely not run it as systemd daemon now, but if we will, it also will be more reliable =). But this is hypothetical at this stage

haircommander replied:
ah I see, done

haircommander committed:
e2e_node: use restart instead of start stop

Signed-off-by: Peter Hunt <pehunt@redhat.com>

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

SergeyKanzhelev commented:
/retest
